patient_data_source.py

```diff
@@ -0,0 +1,214 @@
+from abc import ABC, abstractmethod
+import names
+import warnings
+from datetime import datetime
+from enum import Enum
+from collections.abc import Iterable
+from fhirclient.models.patient import Patient as FHIR_Patient
+from fhirclient.models.observation import Observation as FHIR_Observation
+from fhirclient.models.fhirdate import FHIRDate
+
+class Patient(ABC):
+	class Gender(Enum):
+		male = 0
+		female = 1
+		other = 2
+		unknown = 3
+
+	def get_gender(self) -> Gender:
+		"""Returns the patient's gender. Default implementation return Gender.unkown"""		
+		return Gender.unknown
+
+	def get_identifier_value(self) -> str:
+		"""Returns a id value for external use."""
+		return None
+
+	@abstractmethod
+	def get_dob(self) -> datetime:
+		"""Returns the patient's date of birth."""
+		return datetime.min
+
+	def get_identifier_system(self) -> str:
+		"""Returns the namespace for unique identifier values."""
+		return None
+
+	def generate_name(self, gender : Gender) -> tuple[str, str]:
+		"""Generate a first and last name based on gender."""
+		match gender:
+			case Patient.Gender.male:
+				first_name = names.get_first_name('male')
+			case Patient.Gender.female:
+				first_name = names.get_first_name('female')
+			case default:
+				first_name = names.get_first_name()
+
+		return names.get_last_name(), first_name
+
+	def get_name(self) -> tuple[str, str]:
+		"""Returns a a tuple of the last and first name of the patient. Default implementation generates names based on gender"""
+		return self.generate_name(self.get_gender())
+
+class Observation(ABC):
+	class ObservationType(Enum):
+		fio2 = 0
+		pip = 1
+		peep = 2
+		hr = 3
+		sao2 = 4
+
+	# Units used for ObservationType's valus. Codes that follow the spec in https://ucum.org/ucum.html.
+	UNIT_CODES = {
+	  ObservationType.fio2 : '',
+	  ObservationType.pip : 'cm[H20]',
+	  ObservationType.peep : 'cm[H20]',
+	  ObservationType.hr : '/min',
+	  ObservationType.sao2 : '%',
+	}
+
+	# LOINC codes that I found by using loinc.org/search/ ... and making my best guesses when things were unclear
+	# Not to be fully trusted
+	LOINC_CODES = {
+	  ObservationType.fio2 : '19996-8',
+	  ObservationType.pip : '60951-1',
+	  ObservationType.peep : '20077-4',
+	  ObservationType.hr : '8867-4',
+	  ObservationType.sao2 : '59408-5',
+	}
+
+	# Human readable strings for the ObservationTypes currently implemented.
+	DISPLAY_STRINGS = {
+	  ObservationType.fio2 : 'ETT Sx Quality',
+	  ObservationType.pip : 'PIP',
+	  ObservationType.peep : 'PEEP',
+	  ObservationType.hr : 'Heart Rate',
+	  ObservationType.sao2 : 'Sa02',
+	}
+
+	def get_identifier_value(self) -> str:
+		"""Returns a id value for external use."""
+		return None
+
+	@abstractmethod
+	def get_observation_type(self) -> ObservationType:
+		"""Returns the observation's ObservationType. Used internally for returning other Observation attribues"""
+		pass
+
+	def get_identifier_system(self) -> str:
+		"""Returns the namespace for identifier values."""
```

Comment:
Same comment as above with `Patient.get_identifier_system`-- the terminology "namespace" is not clear